bigdecimal & bigmath

ahorek opened the pull request:
Implemets:
bigdecimal#.to_r
bigmath#.exp

Fixes:
CMP/COERCE bigdecimals -> floats / rationals
Exceptions
Missing require
Tests, clean up

References:
#1186

kares commented:
you seem to have added a lot of new excludes ... those were passing before, right?
... also your commit messages could use some chit-chat :)

ahorek commented:
there are some minor cases when we don't match MRI:
1/ sometimes precision isn't limited
2/ special modes like BigDecimal::ROUND_MODE & BigDecimal::EXCEPTIONS are sometimes ignored / not implemented
3/ BigDecimal.pow works but it's extremely slow (see test_exception_overflow)
4/ MRI's BigDecimal can handle larger exponents, Java's BigInteger class can not
BigDecimal("10E1999999999") MRI => 0.1E2000000000, JRuby => Infinity
5/ #3242

nice to have:
BigMath.log / exp should be implemented natively like MRI did since Ruby 2.0.

all my changes are complatible with MRI 1.9.3, so backporting shouln't be a problem. I've also added comments and fixed broken tests.

enebo commented:
@ahorek can we trouble you for a second PR for jruby-1_7 branch?  I suspect we cannot cherry-pick this.
